packages/yarnpkg-pnpify/sources/buildNodeModulesTree.ts

```diff
@@ -0,0 +1,358 @@
+ * Retrieve full package list and build hoisted `node_modules` directories
+ * representation in-memory.
+ *
+ * @param pnp PnP API
+ *
+ * @returns hoisted `node_modules` directories representation in-memory
+ */
+export const buildNodeModulesTree = (pnp: PnpApi, options: NodeModulesTreeOptions): NodeModulesTree => {
+  const {packageTree, packages, locators} = buildPackageTree(pnp, options);
+
+  const hoistedTree = hoist(packageTree, packages);
+
+  return populateNodeModulesTree(pnp, hoistedTree, locators, options);
+};
+
+/**
+ * Traverses PnP tree and produces input for the `RawHoister`
+ *
+ * @param pnp PnP API
+ *
+ * @returns package tree, packages info and locators
+ */
+const buildPackageTree = (pnp: PnpApi, options: NodeModulesTreeOptions): { packageTree: ReadonlyHoisterPackageTree, packages: HoisterPackageInfo[], locators: PackageLocator[] } => {
+  const packageTree: HoisterPackageTree = [];
+  const packages: HoisterPackageInfo[] = [];
+  const locators: PackageLocator[] = [];
+  const locatorToPackageMap = new Map<LocatorKey, HoisterPackageId>();
+  const packageInfos: PackageInformation<NativePath>[] = [];
+
+  const pnpRoots = pnp.getDependencyTreeRoots();
+
+  let lastPkgId = 0;
+
+  const getLocatorKey = (locator: PackageLocator): LocatorKey => `${locator.name}:${locator.reference}`;
+
+  const assignPackageId = (locator: PackageLocator, pkg: PackageInformation<NativePath>) => {
+    const locatorKey = getLocatorKey(locator);
+    const pkgId = locatorToPackageMap.get(locatorKey);
+    if (typeof pkgId !== 'undefined')
+      return pkgId;
+
+    const newPkgId = lastPkgId++;
+    const packagePath = npath.toPortablePath(pkg.packageLocation);
+    const weight = getPackageWeight(packagePath, options);
+
+    locatorToPackageMap.set(locatorKey, newPkgId);
+    locators.push(locator);
+    packages.push({name: locator.name!, weight});
+    packageInfos.push(pkg);
+    return newPkgId;
+  };
+
+  const addedIds = new Set<HoisterPackageId>();
+  const addPackageToTree = (pkg: PackageInformation<NativePath>, pkgId: HoisterPackageId, parentDepIds: Set<HoisterPackageId>) => {
+    if (addedIds.has(pkgId))
+      return;
+    addedIds.add(pkgId);
+    const deps = new Set<HoisterPackageId>();
+    const peerDeps = new Set<HoisterPackageId>();
+    packageTree[pkgId] = {deps, peerDeps};
+
+    for (const [name, reference] of pkg.packageDependencies) {
+      if (reference !== null) {
+        const locator = typeof reference === 'string' ? {name, reference} : {name: reference[0], reference: reference[1]};
+        const depPkg = pnp.getPackageInformation(locator)!;
+        const depPkgId = assignPackageId(locator, depPkg);
+        if (pkg.packagePeers.has(name)) {
+          peerDeps.add(depPkgId);
+        } else {
+          deps.add(depPkgId);
+        }
+      }
+    }
+
+    const allDepIds = new Set([...deps, ...peerDeps]);
+    for (const depId of allDepIds) {
+      const depPkg = packageInfos[depId];
+      if (depPkg) {
```

Comment:
When is it false? Is it just for TS to be happy (if so I'd prefer an assertion)?